src/ledger/LedgerStateSnapshot.h

```diff
@@ -0,0 +1,179 @@
+class LedgerHeaderWrapper
+{
+    std::variant<LedgerTxnHeader, std::shared_ptr<LedgerHeader>> mHeader;
+
+  public:
+    LedgerHeaderWrapper(LedgerTxnHeader&& header);
+    LedgerHeaderWrapper(LedgerHeader const& header);
+    LedgerHeaderWrapper(std::shared_ptr<LedgerHeader> header);
+    LedgerHeader& currentToModify();
+    LedgerHeader const& current() const;
+    LedgerTxnHeader const&
+    maybeGetLedgerTxnHeader() const
+    {
+        releaseAssert(std::holds_alternative<LedgerTxnHeader>(mHeader));
+        return std::get<0>(mHeader);
+    }
+};
+
+// A unified interface for read-only ledger state snapshot.
+// Supports SQL (via read-only LedgerTxn), as well as BucketList snapshots.
+class AbstractLedgerStateSnapshot
+{
+  public:
+    virtual ~AbstractLedgerStateSnapshot() = default;
+    virtual LedgerHeaderWrapper getLedgerHeader() = 0;
+    virtual LedgerHeaderWrapper getHeader() = 0;
+    virtual LedgerEntryWrapper getSourceAccount(AccountID const& account) = 0;
+    virtual LedgerEntryWrapper
+    getSourceAccount(LedgerHeaderWrapper const& header,
+                     TransactionFrame const& tx) = 0;
+    virtual LedgerEntryWrapper
+    getSourceAccount(LedgerHeaderWrapper const& header,
+                     TransactionFrame const& tx,
+                     AccountID const& AccountID) = 0;
+    virtual LedgerEntryWrapper load(LedgerKey const& key) = 0;
+#ifdef BUILD_TESTS
+    virtual LedgerEntryWrapper loadToModify(LedgerKey const& key) = 0;
+#endif
+};
+
+// A concrete implementation of read-only SQL snapshot wrapper
+class LedgerTxnReadOnly : public AbstractLedgerStateSnapshot
+{
+    // Callers are expected to manage `AbstractLedgerTxn` themselves.
+    // LedgerTxnReadOnly guarantees that LedgerTxn, LedgerTxnHeader and
+    // LedgerTxnEntry referenced all remain valid.
+    AbstractLedgerTxn& mLedgerTxn;
+
+  public:
+    LedgerTxnReadOnly(AbstractLedgerTxn& ltx);
+    ~LedgerTxnReadOnly() override;
+    LedgerHeaderWrapper getLedgerHeader() override;
+    LedgerHeaderWrapper getHeader() override;
+    LedgerEntryWrapper getSourceAccount(AccountID const& account) override;
+    LedgerEntryWrapper getSourceAccount(LedgerHeaderWrapper const& header,
+                                        TransactionFrame const& tx) override;
+    LedgerEntryWrapper getSourceAccount(LedgerHeaderWrapper const& header,
+                                        TransactionFrame const& tx,
+                                        AccountID const& AccountID) override;
+    LedgerEntryWrapper load(LedgerKey const& key) override;
+#ifdef BUILD_TESTS
+    LedgerEntryWrapper loadToModify(LedgerKey const& key) override;
+#endif
+    AbstractLedgerTxn&
+    getLedgerTxn() const
+    {
+        return mLedgerTxn;
+    }
+};
+
+// A concrete implementation of read-only BucketList snapshot wrapper
+class BucketSnapshotState : public AbstractLedgerStateSnapshot
+{
+    BucketManager& mBucketManager;
```

Comment:
We don't need to persist this reference, just take a `BucketManager const&` or `BucketSnapshotManager const&` in the constructor.